### ISSUE TITLE
Import MS/MS spectrum data from various sources

### ISSUE THREAD

**Issue by jorainer** [rformassspectrometry/CompoundDb#23]
Import MS/MS spectrum data from different databases:

- [x] HMDB (http://www.hmdb.ca)
- [x] MoNa (issue #30)
- [ ] Massbank

Seems ChEBI, PubChem and Lipid maps don't provide spectra; can you confirm @stanstrup @michaelwitting ?

**Comment by stanstrup** [comment]
Yes, those don't spectra.

**Comment by jotsetung** [comment]
Little complication from HMDB: HMDB provides one xml file for each spectrum associated with a compound. Now it can be that the same spectrum (same values) are associated to different compounds. HMDB uses the same `spectrum_id`, but provides two (or more) xml files, one for each compound_id.

Complicated solution to handle this would be:
- insert only unique spectra to the `spectrum` table.
- Add an additional table providing the mapping between spectrum and compound tables (to handle the n:m mapping).

Disadvantage: queries are more complicated, possibly slower.

Simple solution:
- insert each spectrum as it is provided, but assigning own, internal and unique, IDs to each spectrum.

**Comment by michaelwitting** [comment]
I have at least a function that can read from MassBank records. Well that doesn't help with MoNA but with all other MassBank related records.
Check my `masstrixR` package later that day. There is a branch called `masstrixR_RaMoNA_merge`. It is based on our in-house tool MassTRIX [1]. There might be also some other usefull functions we can use / reuse.

[1] http://dx.plos.org/10.1371/journal.pone.0039860

**Comment by jotsetung** [comment]
Cool! thanks for your input @michaelwitting . I had a look at the MoNa SDF file and it should be straight forward to extract all relevant information (compound annotations and spectra) from that. It's just a bummer that every database/resource uses own identifiers and nomenclature.

**Comment by michaelwitting** [comment]
Well, this is why I'm working here with the MassBank record format. It is rich in metadata and human readable, but also easy to parse due to partially controlled variables. The functions I wrote are reading from this format to a `Spectra` object and then will also write to a MassBank record.

By the way: I implemented two new spectra comparison methods in `masstrixR`. One is a standards forward score aka dot product, but aligns the spectra instead of binning and the second one is a reverse score (reverse dot product), which uses only peaks that are in the library spectrum. If the match is good both should be quite high, if the forward is low and the backward high then you have a lot "contaminating" peaks in your query spectrum or it is just the wrong hit.

**Comment by jotsetung** [comment]
when you talk MassBank record format - where do you get that data? Is it from https://github.com/MassBank/MassBank-data ? apparently not MoNa...

**Comment by michaelwitting** [comment]
Yes, for example. We use also the MassBank format for our internal database.
Regarding the MoNA JSON: It is very inconsistent. When I read some data from their webservice I'm having difficulties to get the entries I would like to access. Not every library they have has exactly the same format. Maybe it is different with the JSON files...

**Comment by jotsetung** [comment]
Import of open data from MassBank is discussed in issue #34.
